### PR TITLE
Use binary search in level manager

### LINKED ISSUE #58: Use binary search in LevelManager
Binary search rather than scanning can be used in the following places in LevelManager:
* When finding overlap in GetTableIDsInRange - can be used to locate both the segment entry and the table entries
* Finding entry to remove in ApplyDeregistrations - can be used to locate both the segment entry and the table entry
* Finding insert point in ApplyRegistrations - can be used to locate both the segment entry and the table entry
* containsTable method

### PR DESCRIPTION
**Pull request by jlerche**
closes #58

### CODE REVIEW

File: levels/level_manager_test.go [hunk 1]
```diff
@@ -2082,3 +2082,173 @@ func TestRegisterDeadVersionRangeDedup(t *testing.T) {
...
+		tableEntries: []*TableEntry{
+			{
+				SSTableID:  []byte("sstableid1"),
+				RangeStart: []byte("rangestart1"),
+				RangeEnd:   []byte("rangeend1"),
+			},
+			{
+				SSTableID:  []byte("sstableid2"),
+				RangeStart: []byte("rangestart2"),
+				RangeEnd:   []byte("rangeend2"),
```
Comment (purplefox) [inline levels/level_manager_test.go:2097]: The ranges in these two tables are overlapping - that would normally only be the case for L0. I'd suggest testing L0 (overlapping table entries) and L > 0 (no overlapping separately).

File: levels/level_manager.go [hunk 2]
```diff
@@ -1699,12 +1719,89 @@ func (lm *LevelManager) dump() {
 }
 
 func containsTable(seg *segment, tabID sst.SSTableID) bool {
-	for _, te := range seg.tableEntries {
-		if bytes.Equal(te.SSTableID, tabID) {
-			return true
+	n := len(seg.tableEntries)
```
Comment (purplefox) [inline levels/level_manager.go:1722]: I don't think this is right. SSTableIDs are just random UUIDs and won't be in any special order in the segment. I think there's no alternative than iterating over all elements

File: levels/level_manager_test.go [hunk 3]
```diff
@@ -2082,3 +2082,173 @@ func TestRegisterDeadVersionRangeDedup(t *testing.T) {
...
+		{rangeStart: []byte("ggg"), rangeEnd: []byte("iii")},
+	}
+
+	// Test cases
+	testCases := []struct {
+		name         string
+		registration RegistrationEntry
+		expected     int
+	}{
+		{name: "no valid insertion point", registration: RegistrationEntry{KeyStart: []byte("b"), KeyEnd: []byte("e")}, expected: -1},
```
Comment (purplefox) [inline levels/level_manager_test.go:2172]: I think it would be good here to test all the boundary conditions - e.g. when key start / key end are exactly on the range start / end of existing entries, and when they're just one away.

File: levels/level_manager_test.go [hunk 4]
```diff
@@ -2082,3 +2082,173 @@ func TestRegisterDeadVersionRangeDedup(t *testing.T) {
...
+		deRegistration RegistrationEntry
+		expected       int
+	}{
+		{
+			name:           "exact match first segment",
+			segmentEntries: segmentEntries,
+			deRegistration: RegistrationEntry{KeyStart: []byte("aaa"), KeyEnd: []byte("ccc")},
+			expected:       0,
+		},
+		{
```
Comment (purplefox) [inline levels/level_manager_test.go:2208]: as above

**Review by purplefox — CHANGES_REQUESTED** [review]
Looks good! Just a few comments

File: levels/level_manager_test.go (hunk 1 above)
Comment (jlerche) [reply levels/level_manager_test.go:2097]: I ended up deleting this test since the binary search approach for `containsTable` wasn't valid

File: levels/level_manager.go (hunk 2 above)
Comment (jlerche) [reply levels/level_manager.go:1722]: That's right, I'll go ahead and revert it back to the original

File: levels/level_manager_test.go (hunk 3 above)
Comment (jlerche) [reply levels/level_manager_test.go:2099]: I assumed by "one away" you meant in the sense of `[]byte("aaa")` vs `[]byte("aab")`

File: levels/level_manager.go [hunk 5]
```diff
@@ -1707,6 +1727,84 @@ func containsTable(seg *segment, tabID sst.SSTableID) bool {
...
+	return -1
+}
+
+func getSegmentEntryForDeregistration(segmentEntries []segmentEntry, deRegistration RegistrationEntry) int {
+	n := len(segmentEntries)
+	if n == 0 {
+		return -1
+	}
+	index := sort.Search(n, func(i int) bool {
+		return bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) < 0 ||
```
Comment (purplefox) [inline levels/level_manager.go:1765]: Presumably the anonymous function here returns true if there is an overlap between the deregistration range and the segment entry range? If so, there's already a 'hasOverlap' function in this file that does the same and can be reused.

File: levels/level_manager.go [hunk 6]
```diff
@@ -1707,6 +1727,84 @@ func containsTable(seg *segment, tabID sst.SSTableID) bool {
...
+			(bytes.Compare(deRegistration.KeyStart, segmentEntries[i].rangeStart) >= 0 &&
+				bytes.Compare(deRegistration.KeyEnd, segmentEntries[i].rangeEnd) <= 0)
+	})
+	// if index < n it's possible that sort.Search will consider it a legal index
```
Comment (purplefox) [inline levels/level_manager.go:1769]: I thought that sort.Search will return n if not found and something else if found? So not sure why this extra check is necessary. I could be missing something.

File: levels/level_manager.go [hunk 7]
```diff
@@ -1707,6 +1727,84 @@ func containsTable(seg *segment, tabID sst.SSTableID) bool {
...
+				pos = i
+				break
+			}
+		}
+		return pos
+	}
+	n := len(seg.tableEntries)
+	// in principle if the ids are equal their ranges should be as well
+	pos = sort.Search(n, func(i int) bool {
+		return bytes.Compare(seg.tableEntries[i].RangeStart, deRegistration.KeyStart) >= 0
```
Comment (purplefox) [inline levels/level_manager.go:1792]: Shouldn't it be == 0 ?
As we know the correct entry will have exactly matching range

**Review by purplefox — CHANGES_REQUESTED** [review]

File: levels/level_manager.go (hunk 6 above)
Comment (jlerche) [reply levels/level_manager.go:1769]: `sort.Search` only finds the insertion point, so the extra check is required to make sure it actually found the element that's being searched for. The comment I left there is definitely not very clear on that point, now that I read it again.

File: levels/level_manager.go (hunk 7 above)
Comment (jlerche) [reply levels/level_manager.go:1792]: Because `Search` only finds the insertion point, the compare functio needs to partition the boundaries of the search into true/false. This is the implementation of `Search`

```go
```

File: levels/level_manager.go (hunk 5 above)
Comment (jlerche) [reply levels/level_manager.go:1765]: Re-using `hasOverlap` was actually my first thought, but it didn't work out for reasons detailed further below https://github.com/spirit-labs/tektite/pull/180#discussion_r1694283995

File: levels/level_manager.go (hunk 7 above)
Comment (purplefox) [reply levels/level_manager.go:1792]: Sorry if I am being slow.
According to the documentation for search:
```
Search uses binary search to find and return the smallest index i
// in [0, n) at which f(i) is true
```
If we use a function that checks the range exactly using == 0, then it will return the index of the first item where that function is true, which will be the entry we want. What am I missing here?

Comment (jlerche) [reply levels/level_manager.go:1792]: Here's a playground link that shows the difference https://go.dev/play/p/J9CZyYdU04q

the first one uses `>= 1` and we see the `h` values going 5 -> 2 > 1 -> 0, in the second one using `== 1`, we see the `h` values going 5 -> 8 -> 9.

I don't think the documentation is wrong, just not very clear

**Comment by purplefox** [comment]
Thanks for your contribution!
